### DISCUSSION TITLE
Replication suggestion

### DISCUSSION THREAD

**Discussion by VladimirCores** [pocketbase/pocketbase#327] [Polls]
What do you think about replication feature?

**Comment by ganigeorgiev** [comment]
There are already great existing standalone tools for that, like [Lightstream](https://github.com/benbjohnson/litestream) and [LiteFS](https://github.com/superfly/litefs), both by @benbjohnson (I remember reading a quote somewhere by the author explaining the differences that Lightstream was more focused on disaster recovery for a single machine, where LiteFS was intended for running SQLite in a distributed cluster, or something along the lines).

Replication (as in emulating binlog) in general would be a nice to have feature but also a very complex one and unfortunately I don't think I'll have time to work on that (at least for v1.0.0). My recommendation for now is, if you really need it, to explore the above options.

**Reply by ramkumarkb** [reply]
Hi,

Today there was a blog annoucing the beta for LiteFS - https://fly.io/blog/introducing-litefs/ 

I was wondering whether having LiteFS as an option for PocketBase would make sense (yes, it adds additional dependencies like Consul) -  but an additional option could make PB interesting for a broader range of usecases... 

Btw, Thank you so much for this amazing project 👏 💯

**Comment by osseonews** [comment]
Anyone test Lightstream with Pocketbase? Does it have any effect on performance? There are some caveats listed https://litestream.io/tips/

**Reply by benbjohnson** [reply]
Litestream author here. You shouldn't see much performance impact from Litestream. The v0.3.x versions do polling on the SQLite WAL file but it's fairly minimal. The next version (v0.4.0) will use file watchers so its overhead should be even less.

**Reply by rajeshg** [reply]
Given an existing pocketbase application, would be nice to be able to replicate with litefs. 

Looking at the documentation from https://fly.io/docs/litefs/example/. pocketbase on fly.io has it's data setup at /pb. The sqlite db is located at /pb/pb_data/data.db

I am assuming the following fly.toml should be enough. But no luck yet.

```  
app = "pocketbase-cb"
kill_signal = "SIGINT"
kill_timeout = 5
primary_region = "iad"
processes = []

[experimental]
  enable_consul = true

[env]

[experimental]
  auto_rollback = true

[mounts]
  source = "litefs"
  destination = "/pb/pb_data"

[[services]]
  http_checks = []
  internal_port = 8080
  processes = ["app"]
  protocol = "tcp"
  script_checks = []
  [services.concurrency]
    hard_limit = 25
    soft_limit = 20
    type = "connections"

  [[services.ports]]
    force_https = true
    handlers = ["http"]
    port = 80

  [[services.ports]]
    handlers = ["tls", "http"]
    port = 443

  [[services.tcp_checks]]
    grace_period = "1s"
    interval = "15s"
    restart_limit = 0
    timeout = "2s"

```

**Reply by benbjohnson** [reply]
The `[mounts]` section simply mounts a regular volume in `fly.toml`. You'll need to configure LiteFS using a `litefs.yml` and then run `litefs mount` as part of your `Dockerfile`'s `ENTRYPOINT` / `CMD`.

The upcoming release of LiteFS v0.4.0 will also have an optional thin proxy layer that should make hosting regular SQLite web apps a lot simpler. It handles write forwarding & read-your-writes on replicas automatically. I'll try write up an example of PocketBase running on LiteFS v0.4.0 once it's released.

**Reply by ndarilek** [reply]
I've been looking into this today and it seems potentially doable, but it occurs to me that events may be entirely in-process so you'd probably lose lots of functionality (E.g. reacting to model changes, real-time subscriptions, etc.) by running on LiteFS.

Is that accurate? Or is there some lower-level change-monitoring mechanism that lets events work across processes as long as they're accessing the same database?

**Reply by ganigeorgiev** [reply]
@ndarilek App event hooks and realtime subscription events are triggered on app level and are not stored anywhere. Currently there is no builtin mechanism to sync or propagate those events across multiple PocketBase instances (if that's what you are asking for).

**Reply by ndarilek** [reply]
Gotcha, yeah, that definitely puts a wrinkle in things.

How bad of an idea would it be to persist events to a table that non-primaries can read from to dispatch the events in their own processes, maybe with some cleanup logic for stale events? Or is there a better way?

Not sure if this would actually get us anything, I'm mainly interested in seeing how far we can push this.

**Reply by ganigeorgiev** [reply]
Yes, that's one way to do it, but this may not work with all events and the behavior for some would be ambiguous as the event hook chain will be broken (eg. instance A could decide to stop the propagation of a certain event or return an error in the middle of the execution, while other instances will not be notified about this; or what if you want to change the response body in a request hook from an instance that it is not actually serving the request, etc.).

It is certainly doable if there are clearly defined rules when and what actually get synced and called, but at least for now this remains out of scope for v1.

If plain vertical scaling is not an option for your expected load (_there will be testbed repo and benchmark tests with the next release_), then PocketBase may get in the way and it could be easier to test other platforms designed with that in mind like supabase, nhost, surrealdb, edgedb, etc. or write your own tailored solution from scratch.

**Reply by de1mat** [reply]
> The upcoming release of LiteFS v0.4.0 will also have an optional thin proxy layer that should make hosting regular SQLite web apps a lot simpler. It handles write forwarding & read-your-writes on replicas automatically. I'll try write up an example of PocketBase running on LiteFS v0.4.0 once it's released.

Really interested in PocketBase running on LiteFS, looking forward to the example Ben.

**Reply by benbjohnson** [reply]
@de1mat From my understanding, it sounds like there's in-memory state for Pocketbase that we would be unable to replicate easily. We do have a feature for propagating events between nodes (https://github.com/superfly/litefs/issues/20) so that could be a possibility once we implement that but it would require some support within Pocketbase itself to utilize it.

**Reply by khrome83** [reply]
Out of curiosity, there are memory based events. What features of Pocketbase.io actually trigger events?

I understand real time. I don't use that. But what other things trigger events, outside of using Go and extending it as a framework, or adding webhooks.

Trying to understand, if I use LiteFS, what features are actually impacted.

I am also curious if it's a request/activity that would trigger a event, if it could be hacked to use `fly-replay` header?
And could webhooks, and other events be limited to some type of identifier that it is primary?

I know this is specific to fly.io and LiteFS. For my use case, i am trying to use Pocketbase as a BaaS for managing data. And the website uses read only copies of the database for server side access of Sqlite for rendering data. No need for JS and client side integrations then.

**Reply by ganigeorgiev** [reply]
@khrome83

> What features of Pocketbase.io actually trigger events?

Events are triggered for various operations - modifying collection records, sending emails, when accessing API routes, etc.  Practically for anything listed in https://pocketbase.io/docs/go-event-hooks/ (_all event hooks are triggered at some point internally from the application_).

If you don't have custom app hooks listeners, don't care about the realtime subscriptions and the replicas are read-only, then in this case I guess the hook events will not affect you and LiteFS may work for your specific use case.

**Reply by khrome83** [reply]
Thank you for the quick response and your dedication to answering questions. Your giving better support than VC funded projects with full teams.... just saying.

**Comment by MedLabs** [comment]
I would like to see a pouchdb-like replication feature in the js-sdk

**Comment by denzuko** [comment]
IMHO replication is a hard-ish requirement for our usage at 3umgroup and Daplanet. However instead of SQLite perhaps extending the driver system to support postgreql or MySQL would be a better fit?

The idea there being SQLite fits most users while other that need replication would already have that baked in to their chosen database engine.

**Comment by VladimirCores** [comment]
We are looking at (and Im personally will try) https://electric-sql.com/ and "local first" approach for some projects.
